[SelectionDAG] Let ComputeKnownSignBits handle (shl (ext X), C)

bjope opened the pull request:
Add simple support for looking through ZEXT/ANYEXT/SEXT when doing ComputeKnownSignBits for SHL. This is valid for the case when all extended bits are shifted out, because then the number of sign bits can be found by analysing the EXT operand.

A future improvement could be to pass along the "shifted left by" information in the recursive calls to ComputeKnownSignBits. Allowing us to handle this more generically.

bjope committed:
[SelectionDAG] Let ComputeKnownSignBits handle (shl (ext X), C) (#97695)

Add simple support for looking through ZEXT/ANYEXT/SEXT when doing
ComputeKnownSignBits for SHL. This is valid for the case when all
extended bits are shifted out, because then the number of sign bits
can be found by analysing the EXT operand.

A future improvement could be to pass along the "shifted left by"
information in the recursive calls to ComputeKnownSignBits. Allowing
us to handle this more generically.

bjope commented:
No regression tests yet. If anyone has a good idea how and where to put them, let me know.

The need for this was found downstream as DAG Combiner prefer `(SHL (ANY_EXTEND (ASHR ...)))` instead of `(SHL (SIGN_EXTEND (ASHR ...)))`, which actually is bad for ComputeNumSignBits.

jayfoad commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:
```diff
@@ -4617,6 +4617,21 @@ unsigned SelectionDAG::ComputeNumSignBits(SDValue Op, const APInt &DemandedElts,
+        EVT ExtendedOpVT = ExtendedOp.getValueType();
+        uint64_t ExtendedWidth =
+            SrcVT.getScalarSizeInBits() - ExtendedOpVT.getScalarSizeInBits();
+        if (ExtendedWidth <= *ShAmt) {
```
This seems wrong in the vector case, where the shift amounts could be different and *ShAmt is the max shift amount. In that case, some of the other shift amounts could be < ExtendedWidth.

bjope replied:
Indeed it was wrong, thanks!
I've added a fixup to check for MimimumShiftAmount.

jayfoad replied:
Looks better, thanks.

bjope committed:
[CodeGen] Pre-commit test case related to ComputeNumSignBits for SHL (#97695)

Adding test cases aiming at showing possibility to look through
ZERO_EXTEND/ANY_EXTEND when computing number of sign bits for an
SHL node. If all extended bits are shifted out we can analyze the
operand that is extended.

bjope commented:
Added some test cases based on the `(sshlsat x, c) -> (shl x, c)` fold to verify number of sign bits that are computed.

RKSimon commented on llvm/test/CodeGen/X86/computenumsignbits-shl.ll:
```diff
@@ -0,0 +1,158 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
+; RUN: llc < %s -mtriple=x86_64-linux | FileCheck %s --check-prefix=X64
```
(pedantic) rename to `known-signtbits-shl.ll` to

RKSimon commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:
```diff
@@ -4615,12 +4615,30 @@ unsigned SelectionDAG::ComputeNumSignBits(SDValue Op, const APInt &DemandedElts,
             getValidMaximumShiftAmount(Op, DemandedElts, Depth + 1)) {
+      if (Op.getOperand(0).getOpcode() == ISD::ANY_EXTEND ||
+          Op.getOperand(0).getOpcode() == ISD::ZERO_EXTEND ||
+          Op.getOperand(0).getOpcode() == ISD::SIGN_EXTEND)
```
Use ISD::isExtOpcode(Op.getOperand(0).getOpcode())

RKSimon commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:
```diff
@@ -4615,12 +4615,30 @@ unsigned SelectionDAG::ComputeNumSignBits(SDValue Op, const APInt &DemandedElts,
+          SDValue Src = Op.getOperand(0);
+          EVT SrcVT = Src.getValueType();
+          SDValue ExtendedOp = Op.getOperand(0).getOperand(0);
+          EVT ExtendedOpVT = ExtendedOp.getValueType();
```
(pedantic) ExtendedOp isn't extended - it's the pre-extended source value

RKSimon reviewed (commented):
A couple of minor comments

RKSimon commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:
```diff
@@ -4615,12 +4615,28 @@ unsigned SelectionDAG::ComputeNumSignBits(SDValue Op, const APInt &DemandedElts,
+        EVT ExtendeeVT = Extendee.getValueType();
+        uint64_t SizeDifference =
+            ExtVT.getScalarSizeInBits() - ExtendeeVT.getScalarSizeInBits();
+        if (SizeDifference <= MinShAmt) {
```
Add a explanation comment (similar to the patch summary)

RKSimon reviewed (approved):
LGTM - cheers

bjope commented:
> LGTM - cheers

Thanks!
